doc/monads/readers.lean

```diff
@@ -0,0 +1,196 @@
+recommended style.  You will see this patten used heavily in Lean code.
+
+The `let env ← read` expression in `readerFunc1` unwraps the environment from the `ReaderM` so we
+can use it. Each type of monad might provide one or more extra functions like this, functions that
+become available only when you are in the context of that monad.
+
+Here the `readerFunc2` function uses the `bind` operator `>>=` just to show you that there are bind
+operations happening here.  The `readerFunc3` function uses the  `do` notation you learned about in
+[Monads](monads.lean.md) which hides that bind operation and can make the code look cleaner.
+
+The `do` notation with `let x ← readerFunc2` is also calling the `run` function to actually execute
+the monadic action, then it unwraps the value x, so it is really doing `let x ← readerFunc2.run`.
+
+The important difference here to the earlier code is that `readerFunc3` and `readerFunc2` no longer
+have an **explicit** Environment input parameter that needs to be passed along all the way to
+`readerFunc1`.  Instead, the `ReaderM` monad is taking care of that for you, which gives you the
+illusion of something like global state where the state is now available to all functions that use
+the `ReaderM` monad.
+
+The above code also introduces an important idea. Whenever you learn about a monad "X", there's
+often (but not always) a `run` function to execute that monad, and sometimes some additional
+functions like `read` that interact with the monad state.
+
+You might be wondering, how does the state actually move through the `ReaderM` monad? How can you
+add an input argument to a function by modifying it's return type?  There is a special command in
+the Lean interpreter that will show you the reduced Types:
+-/
+#reduce ReaderM Environment String   -- Environment → String
+/-!
+And you can see here that this type is actually a function!  It's a function that takes an
+`Environment` as input and returns a `String`.  So, remember in Lean that a function that takes
+argument a and returns a string: `def f (a : Nat) → String` is the same as the function that takes no
+arguments and returns another function of type `Nat → String`.  Well this fact is being used by the
+ReaderM Monad to add an input argument to all the functions that use the `ReaderM` monad and this is why
+`main` is able to start things off by simply passing that new input argument in `readerFunc3 env`.
+So now that you know the implementation details of the `ReaderM` monad you can see that what it is
+doing looks very much like the original code we wrote at the beginning of this section, only it's
+taking a lot of the tedious work off your plate and it is creating a nice clean separation between
+what your pure functions are doing, and the global state idea that the `ReaderM` adds.
+
+## withReader
+
+One `ReaderM` function can call another with a modified version of the `ReaderM` state.
```

Comment:
This section needs to be improved. The `withReader` is poorly motivated. We should say that `withReader f act` executes `act` in context modified by `f`. We should emphasize that the modification affects only `act`. Adding an example would help too.

Reply:
How's this instead?

## withReader

One `ReaderM` function can call another with a modified version of the `ReaderM` context.
You can use the `withReader` function from the `MonadWithReader` typeclass to do this:

-/
def readerFunc3WithReader : ReaderM Environment String := do
  let f := (λ env => { env with user := "new user" })
  let x ← withReader f readerFunc2
  return "Result: " ++ (toString x)

/-!

So `withReader f m` executes monad `m` in the `ReaderM` context modified by `f`.

Reply:
Fixed in https://github.com/leanprover/lean4/pull/1505/commits/8dd24bc9dae40eb19b4625bf0b94cfb3b15a2f07